File: scripts/checkLockFiles.ts
import { existsSync } from 'node:fs';

const checkLockFiles = (): string[] => {
  const uesYarnLockFileMessage = 'Please remove it and use only "pnpm-lock.yaml".';
  const invalidFileMessage = (lockFile: string): string =>
    `Invalid occurrence of "${lockFile}" file. ${uesYarnLockFileMessage}`;

  const errors: string[] = [];

  const npmLockFile = 'package-lock.json';
  existsSync(npmLockFile) && errors.push(invalidFileMessage(npmLockFile));

  const yarnLockFile = 'yarn.lock';
  existsSync(yarnLockFile) && errors.push(invalidFileMessage(yarnLockFile));

  const pnpmLockFile = 'pnpm-lock.yaml';
  !existsSync(pnpmLockFile) &&
    errors.push('The "pnpm-lock.yaml" does not exist or cannot be read. Please run "pnpm i".');

  return errors;
};

const invalidLockFileErrors = checkLockFiles();

if (invalidLockFileErrors.length > 0) {
  for (const error of invalidLockFileErrors) console.log(error);
  process.exit(1);
}

process.exit(0);

File: tools/src/check_lock_files.go
package main

import (
	"fmt"
	"os"
)

var errors []string

func main() {
	fmt.Println("ℹ️ Checking correct usage of only pnpm lock file ...")

	fileExists("package-lock.json")
	fileExists("yarn.lock")
	fileNotExists("pnpm-lock.yaml")

	if len(errors) > 0 {
		for _, err := range errors {
			fmt.Println(err)
		}
		os.Exit(1)
	}

	fmt.Println("✅ Lock files are looking good.")

	os.Exit(0)
}

func fileExists(fileName string) {
	if _, err := os.Stat(fileName); !os.IsNotExist(err) {
		errors = append(errors, fmt.Sprintf("Invalid occurrence of \"%s\" file. Please remove it and use only \"pnpm-lock.yaml\".", fileName))
	}
}

func fileNotExists(fileName string) {
	if _, err := os.Stat(fileName); os.IsNotExist(err) {
		errors = append(errors, fmt.Sprintf("The \"%s\" does not exist or cannot be read. Please run \"pnpm i\".", fileName))
	}
}
